### ISSUE TITLE
Allow full deallocation of GPU memory

### ISSUE THREAD

**Issue by dtegunov** [tensorflow/tensorflow#15880]
When using the TF C++ library inside an application that also uses GPUs for other tasks (not implemented in TF), it would be useful to be able to deallocate all the GPU memory TF has allocated once the session is closed, and no further TF calls are expected for the time being. gpu_options.allow_growth keeps TF's allocated pool small, but it still can grow to several GB. Even after the session is deleted, the pool doesn't shrink. To free it up, the whole application must be restarted, if I'm not mistaken.

Being able to destroy the [ProcessState](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/process_state.cc) singleton seems to solve it without breaking anything. However, its destructor is protected. Alternatively, getting the Allocator for each GPU from ProcessState and manually destroying them does the trick, but renders TF unusable for all future operations because ProcessState still thinks the Allocators exist and doesn't recreate them when they are required again.

I think making the ProcessState destructor public (or adding a public method to invoke similar code) would be the best solution, but maybe I'm missing an obvious solution that already exists?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by balajichandramohan** [comment]
@dtegunov I am also facing the same issue, I want to run multiple iteration and each iteration is model generation. Can you please provide your code to deallocate GPU memory by manually destroying them?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by asimshankar** [comment]
@zheng-xq may be in a better position to comment on the pros-and-cons of allowing `ProcessState` destruction or a backdoor to releasing memory.

**Comment by zheng-xq** [comment]
@asimshankar The reason the memory belongs to ProcessState even after session is closed is because Variables lives longer than session. But with "reset", all variables and other resources are also released.

So for DirectSession, we can release all memory if the "reset" is called. If that is already done that way, we can certainly add that functionality.

What do you think?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by balajichandramohan** [comment]
From my understand the TensorFlow memory allocation is tied to the process state, given that as a fix I spawn sub process and ran the model generation as part of it. Once the subprocess is done, memory is clear out.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 22 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 37 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 52 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 67 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 82 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 97 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 112 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 127 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @zheng-xq: It has been 142 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @smit-hinsu, @azaks2: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @smit-hinsu, @azaks2: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @smit-hinsu, @azaks2: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @smit-hinsu, @azaks2: It has been 61 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by pwais** [comment]
@tensorflowbutler see for example https://github.com/tensorflow/tensorflow/issues/20387

**Comment by dtegunov** [comment]
Since other users are still looking for a solution, here is the workaround I've been using in my projects:

The trick is to make the destructor in tensorflow/core/common_runtime/process_state.h public instead of protected. Now you can call tensorflow::ProcessState::singleton()->~ProcessState() to free all memory. TF 1.10 is the latest version I've tested for this. Here is a fork with the code already modified: https://github.com/dtegunov/tensorflow_1.10_windows

Make sure to destroy any TF objects that rely on GPU memory prior to doing this, as there is no mechanism to make them aware of the deallocation. After calling the destructor, you can create new TF objects without any issues.

This solution is only guaranteed to work if you can ship a custom TF binary with your product. With the official builds, I think you're running into undefined behavior in C++ when forcibly calling a method that wasn't marked public at compile time.

If performance is not a priority, you can also look into different allocation mechanisms offered in TF. You can force it to call cudaMalloc/cudaFree for every allocation/deallocation instead of grabbing a large chunk and holding it indefinitely. However, this will be much slower than the default allocator. See tensorflow/core/common_runtime/gpu/gpu_process_state.cc for details.

**Comment by smit-hinsu** [comment]
Note that Sessions are deprecated with TensorFlow 2.0 so this shouldn't be an issue for training.
https://medium.com/tensorflow/effective-tensorflow-2-0-best-practices-and-whats-changed-a0ca48767aff

However, Sessions are also useful for inference in C++ so I will keep this open and will look into deallocating memory on session destruction at a lower priority.

**Comment by p890040** [comment]
Very Hope tensorflow official can provide the release gpu memory function, or just make it release after end this thread (not process).

**Comment by EKami** [comment]
I'm also encountering this issue with TF 1.13.1 in eager mode so it's not only TF session related. So far I'm out of luck for finding a solution =/

**Comment by zheng-xing** [comment]
> Since other users are still looking for a solution, here is the workaround I've been using in my projects:
> 
> The trick is to make the destructor in tensorflow/core/common_runtime/process_state.h public instead of protected. Now you can call tensorflow::ProcessState::singleton()->~ProcessState() to free all memory. TF 1.10 is the latest version I've tested for this. Here is a fork with the code already modified: https://github.com/dtegunov/tensorflow_1.10_windows
> 
> Make sure to destroy any TF objects that rely on GPU memory prior to doing this, as there is no mechanism to make them aware of the deallocation. After calling the destructor, you can create new TF objects without any issues.
> 
> This solution is only guaranteed to work if you can ship a custom TF binary with your product. With the official builds, I think you're running into undefined behavior in C++ when forcibly calling a method that wasn't marked public at compile time.
> 
> If performance is not a priority, you can also look into different allocation mechanisms offered in TF. You can force it to call cudaMalloc/cudaFree for every allocation/deallocation instead of grabbing a large chunk and holding it indefinitely. However, this will be much slower than the default allocator. See tensorflow/core/common_runtime/gpu/gpu_process_state.cc for details.

Hi @dtegunov thanks for your suggesting. Can you expand more on "Make sure to destroy any TF objects that rely on GPU memory prior to doing this"? How to destroy TF objects? Thanks in advance.

**Comment by dtegunov** [comment]
TensorFlow's C API has a method called TF_DeleteTensor if I remember correctly, and probably similar methods for other object types that can consume GPU memory. If you don't destroy the objects, they will still have pointers to memory that has been freed through ~ProcessState(), and will fail the next time they try to access that memory.

**Comment by atztao** [comment]
Same issue, need to solve.

**Comment by mohantym** [comment]
Hi @dtegunov ! Have you checked this[ thread](https://www.py4u.net/discuss/17174) on deallocating Gpu memory yet?

**Comment by sullivak-kyra** [comment]
@mohantym the proposed workarounds in that thread are for python, @dtegunov asked about C++.

**Comment by Venkat6871** [comment]
Hi,
Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow version with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the release notes to stay up to date with the latest developments which are happening in the Tensorflow space.

Thank you!
